Sakshi Jha 00211192024 AIML Group- H Multimodal data analysis

**Experiment 2**

**Aim**

To perform text normalization on social media text by applying lowercasing, punctuation handling, contraction expansion, and simple handling of numbers and emojis.

**Objectives**
To understand the need for text normalization.
To convert text into lowercase.
To handle punctuation and special characters.
To expand common contractions such as don't → do not.
To handle numbers and emojis using simple rules.
To compare the original text with the normalized text.

**Requirements** Pandas,
Regular Expression (re) and
NLTK

**DATASET:** Twitter Tweets Sentiment Dataset

Text data: Tweets [Link](https://www.kaggle.com/datasets/yasserh/twitter-tweets-sentiment-dataset/data)


Dataset details:

*   Dataset name : Twitter Tweets Sentiment Dataset
*   Source : Kaggle
*   Data type : Text/ Social Media Data
*   Category : Sentiment analysis
*   Main text attribute: text
*   Label attribute : airline_sentiment
*   Modality : Text










**Theory:**
Text normalization is the process of converting raw text into a more consistent and standard form before further NLP tasks..
For example : social media text often contains - uppercase or lowercase letters, numbers and emojis.

Text normlization make the text easier to process and analyze.

Importing libraries and dataset

In [1]:
import pandas as pd
import re
import nltk

print("Libraries imported successfully")

Libraries imported successfully


In [3]:
# Load the Dataset
url = "https://raw.githubusercontent.com/satyajeetkrjha/kaggle-Twitter-US-Airline-Sentiment-/master/Tweets.csv"

df = pd.read_csv(url)

print("Dataset loaded successfully.")
print("Dataset Shape:", df.shape)

Dataset loaded successfully.
Dataset Shape: (14640, 15)


- Dataset Size: 14,640 rows and 15 attributes
- Each row represents one tweet related to an airline.

In [4]:
# check dataset
print("Column Names:")
print(df.columns.tolist())

print("\nFirst 5 Records:")
df.head()

Column Names:
['tweet_id', 'airline_sentiment', 'airline_sentiment_confidence', 'negativereason', 'negativereason_confidence', 'airline', 'airline_sentiment_gold', 'name', 'negativereason_gold', 'retweet_count', 'text', 'tweet_coord', 'tweet_created', 'tweet_location', 'user_timezone']

First 5 Records:


,tweet_id,airline_sentiment,airline_sentiment_confidence,negativereason,negativereason_confidence,airline,airline_sentiment_gold,name,negativereason_gold,retweet_count,text,tweet_coord,tweet_created,tweet_location,user_timezone
0,570306133677760513,neutral,1.0000,NaN,NaN,Virgin America,NaN,cairdin,NaN,0,@VirginAmerica What @dhepburn said.,NaN,2015-02-24 11:35:52 -0800,NaN,Eastern Time (US & Canada)
1,570301130888122368,positive,0.3486,NaN,0.0000,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica plus you've added commercials t...,NaN,2015-02-24 11:15:59 -0800,NaN,Pacific Time (US & Canada)
2,570301083672813571,neutral,0.6837,NaN,NaN,Virgin America,NaN,yvonnalynn,NaN,0,@VirginAmerica I didn't today... Must mean I n...,NaN,2015-02-24 11:15:48 -0800,Lets Play,Central Time (US & Canada)
3,570301031407624196,negative,1.0000,Bad Flight,0.7033,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica it's really aggressive to blast...,NaN,2015-02-24 11:15:36 -0800,NaN,Pacific Time (US & Canada)
4,570300817074462722,negative,1.0000,Can't Tell,1.0000,Virgin America,NaN,jnardino,NaN,0,@VirginAmerica and it's a really big bad thing...,NaN,2015-02-24 11:14:45 -0800,NaN,Pacific Time (US & Canada)


In [5]:
# check text and column
print("Text Column:")
print(df["text"].head())

print("\nSentiment Column:")
print(df["airline_sentiment"].head())

Text Column:
0                  @VirginAmerica What @dhepburn said.
1    @VirginAmerica plus you've added commercials t...
2    @VirginAmerica I didn't today... Must mean I n...
3    @VirginAmerica it's really aggressive to blast...
4    @VirginAmerica and it's a really big bad thing...
Name: text, dtype: object

Sentiment Column:
0     neutral
1    positive
2     neutral
3    negative
4    negative
Name: airline_sentiment, dtype: object


The dataset contains 15 attributes. but the main focuses on text normalization, the text attribute is selected as the main attribute.

text contains the tweet text that will be normalized.


text → This is the actual tweet on which we perform normalization.

airline_sentiment → useful as the label, but not required for normalization itself.

airline → not required for the normalization task.

retweet_count → not required for the normalization task.

In [6]:
df2 = df[["text"]].copy()

print("Selected Attribute:")
print(df2.columns.tolist())

print("Dataset Shape:", df2.shape)

Selected Attribute:
['text']
Dataset Shape: (14640, 1)


Checking Missing values in the selected text attribute are checked before starting the normalization process.

In [7]:
print("Missing values:")
print(df2.isnull().sum())

Missing values:
text    0
dtype: int64


Select Sample
The dataset contains 14,640 tweets. A sample of 100 tweets is selected for normalization.

In [8]:
df_sample = df2.head(100).copy()

print("Number of tweets selected:", len(df_sample))
print("Sample Shape:", df_sample.shape)

Number of tweets selected: 100
Sample Shape: (100, 1)


Display Raw Tweets: It is important for compare before and after

In [9]:
print("Original Tweets:")
df_sample["text"].head(10)

Original Tweets:


,text
0,@VirginAmerica What @dhepburn said.
1,@VirginAmerica plus you've added commercials t...
2,@VirginAmerica I didn't today... Must mean I n...
3,@VirginAmerica it's really aggressive to blast...
4,@VirginAmerica and it's a really big bad thing...
5,@VirginAmerica seriously would pay $30 a fligh...
6,"@VirginAmerica yes, nearly every time I fly VX..."
7,@VirginAmerica Really missed a prime opportuni...
8,"@virginamerica Well, I didn't…but NOW I DO! :-D"
9,"@VirginAmerica it was amazing, and arrived an ..."


The raw tweets clearly contain the things that is need for this experiment:

Uppercase letters → What

Mentions → @VirginAmerica

Contractions → didn't, it's

Punctuation → ., !, ...

Emoji/emoticon → :-D

**Lowercasing**

All alphabetic characters in the tweets are converted to lowercase to maintain a consistent text format.

In [10]:
df_sample["lowercase_text"] = df_sample["text"].str.lower()

print("Before and After Lowercasing:")
display(df_sample[["text", "lowercase_text"]].head(10))

Before and After Lowercasing:


,text,lowercase_text
0,@VirginAmerica What @dhepburn said.,@virginamerica what @dhepburn said.
1,@VirginAmerica plus you've added commercials t...,@virginamerica plus you've added commercials t...
2,@VirginAmerica I didn't today... Must mean I n...,@virginamerica i didn't today... must mean i n...
3,@VirginAmerica it's really aggressive to blast...,@virginamerica it's really aggressive to blast...
4,@VirginAmerica and it's a really big bad thing...,@virginamerica and it's a really big bad thing...
5,@VirginAmerica seriously would pay $30 a fligh...,@virginamerica seriously would pay $30 a fligh...
6,"@VirginAmerica yes, nearly every time I fly VX...","@virginamerica yes, nearly every time i fly vx..."
7,@VirginAmerica Really missed a prime opportuni...,@virginamerica really missed a prime opportuni...
8,"@virginamerica Well, I didn't…but NOW I DO! :-D","@virginamerica well, i didn't…but now i do! :-d"
9,"@VirginAmerica it was amazing, and arrived an ...","@virginamerica it was amazing, and arrived an ..."


In [18]:
contractions = {
    "don't": "do not",
    "didn't": "did not",
    "it's": "it is",
    "isn't": "is not",
    "can't": "cannot",
    "won't": "will not",
    "you've": "you have",
    "I've": "I have",
    "I'm": "I am",
    "we're": "we are",
    "they're": "they are"
}

def expand_contractions(text):
    for contraction, expanded in contractions.items():
        text = re.sub(
            r"\b" + re.escape(contraction) + r"\b",
            expanded,
            text,
            flags=re.IGNORECASE
        )
    return text

df_sample["expanded_text"] = df_sample["lowercase_text"].apply(expand_contractions)

print("Before and After Contraction Expansion:")
display(df_sample[["lowercase_text", "expanded_text"]].head(10))

Before and After Contraction Expansion:


,lowercase_text,expanded_text
0,@virginamerica what @dhepburn said.,@virginamerica what @dhepburn said.
1,@virginamerica plus you've added commercials t...,@virginamerica plus you have added commercials...
2,@virginamerica i didn't today... must mean i n...,@virginamerica i did not today... must mean i ...
3,@virginamerica it's really aggressive to blast...,@virginamerica it is really aggressive to blas...
4,@virginamerica and it's a really big bad thing...,@virginamerica and it is a really big bad thin...
5,@virginamerica seriously would pay $30 a fligh...,@virginamerica seriously would pay $30 a fligh...
6,"@virginamerica yes, nearly every time i fly vx...","@virginamerica yes, nearly every time i fly vx..."
7,@virginamerica really missed a prime opportuni...,@virginamerica really missed a prime opportuni...
8,"@virginamerica well, i didn't…but now i do! :-d","@virginamerica well, i did not…but now i do! :-d"
9,"@virginamerica it was amazing, and arrived an ...","@virginamerica it was amazing, and arrived an ..."


**Contraction Expansion**
Common contractions in the tweets are expanded into their complete forms.
Examples:
- don't → do not
- didn't → did not


**Punctuation Handling**
Punctuation marks are removed from the lowercase text to create a simpler and more consistent text format.

In [19]:
df_sample["no_punctuation_text"] = df_sample["lowercase_text"].apply(
    lambda x: re.sub(r"[^\w\s]", "", x)
)

print("Before and After Punctuation Handling:")
display(df_sample[["lowercase_text", "no_punctuation_text"]].head(10))

Before and After Punctuation Handling:


,lowercase_text,no_punctuation_text
0,@virginamerica what @dhepburn said.,virginamerica what dhepburn said
1,@virginamerica plus you've added commercials t...,virginamerica plus youve added commercials to ...
2,@virginamerica i didn't today... must mean i n...,virginamerica i didnt today must mean i need t...
3,@virginamerica it's really aggressive to blast...,virginamerica its really aggressive to blast o...
4,@virginamerica and it's a really big bad thing...,virginamerica and its a really big bad thing a...
5,@virginamerica seriously would pay $30 a fligh...,virginamerica seriously would pay 30 a flight ...
6,"@virginamerica yes, nearly every time i fly vx...",virginamerica yes nearly every time i fly vx t...
7,@virginamerica really missed a prime opportuni...,virginamerica really missed a prime opportunit...
8,"@virginamerica well, i didn't…but now i do! :-d",virginamerica well i didntbut now i do d
9,"@virginamerica it was amazing, and arrived an ...",virginamerica it was amazing and arrived an ho...


**Note:** In the final normalization, contractions are expanded before punctuation is removed so that forms such as "didn't" can be correctly converted to "did not".

**Handle Numbers and Emojis**
Numbers and emojis are removed using simple regular expression rules to obtain cleaner text.

example:"I paid $30 😊" → "I paid"

In [22]:
def handle_numbers_emojis(text):
    text = re.sub(r"\d+", "", text)
    text = re.sub(r"[^\w\s]", " ", text)
    text = re.sub(r"\s+", " ", text).strip()
    return text

df_sample["normalized_text"] = df_sample["expanded_text"].apply(
    handle_numbers_emojis
)

print("Before and After Handling Numbers and Emojis:")
display(df_sample[["expanded_text", "normalized_text"]].head(10))

Before and After Handling Numbers and Emojis:


,expanded_text,normalized_text
0,@virginamerica what @dhepburn said.,virginamerica what dhepburn said
1,@virginamerica plus you have added commercials...,virginamerica plus you have added commercials ...
2,@virginamerica i did not today... must mean i ...,virginamerica i did not today must mean i need...
3,@virginamerica it is really aggressive to blas...,virginamerica it is really aggressive to blast...
4,@virginamerica and it is a really big bad thin...,virginamerica and it is a really big bad thing...
5,@virginamerica seriously would pay $30 a fligh...,virginamerica seriously would pay a flight for...
6,"@virginamerica yes, nearly every time i fly vx...",virginamerica yes nearly every time i fly vx t...
7,@virginamerica really missed a prime opportuni...,virginamerica really missed a prime opportunit...
8,"@virginamerica well, i did not…but now i do! :-d",virginamerica well i did not but now i do d
9,"@virginamerica it was amazing, and arrived an ...",virginamerica it was amazing and arrived an ho...


**Compare Before and After Normalization**

The original tweets are compared with the normalized tweets to observe the changes made during the normalization process.

In [23]:
comparison = df_sample[["text", "normalized_text"]].head(10).copy()

comparison.columns = ["Original Text", "Normalized Text"]

display(comparison)

,Original Text,Normalized Text
0,@VirginAmerica What @dhepburn said.,virginamerica what dhepburn said
1,@VirginAmerica plus you've added commercials t...,virginamerica plus you have added commercials ...
2,@VirginAmerica I didn't today... Must mean I n...,virginamerica i did not today must mean i need...
3,@VirginAmerica it's really aggressive to blast...,virginamerica it is really aggressive to blast...
4,@VirginAmerica and it's a really big bad thing...,virginamerica and it is a really big bad thing...
5,@VirginAmerica seriously would pay $30 a fligh...,virginamerica seriously would pay a flight for...
6,"@VirginAmerica yes, nearly every time I fly VX...",virginamerica yes nearly every time i fly vx t...
7,@VirginAmerica Really missed a prime opportuni...,virginamerica really missed a prime opportunit...
8,"@virginamerica Well, I didn't…but NOW I DO! :-D",virginamerica well i did not but now i do d
9,"@VirginAmerica it was amazing, and arrived an ...",virginamerica it was amazing and arrived an ho...


Normalization Stages
The different stages of normalization are displayed for one tweet to understand how the text changes.

In [15]:
sample_index = 2

print("Original Text:")
print(df_sample.loc[sample_index, "text"])

print("\nAfter Lowercasing:")
print(df_sample.loc[sample_index, "lowercase_text"])

print("\nAfter Contraction Expansion:")
print(df_sample.loc[sample_index, "expanded_text"])

print("\nFinal Normalized Text:")
print(df_sample.loc[sample_index, "normalized_text"])

Original Text:
@VirginAmerica I didn't today... Must mean I need to take another trip!

After Lowercasing:
@virginamerica i didn't today... must mean i need to take another trip!

After Contraction Expansion:
@virginamerica i did not today... must mean i need to take another trip!

Final Normalized Text:
virginamerica i did not today must mean i need to take another trip


**Observation**

- 100 tweets were selected from the dataset.
- Lowercasing made the text consistent.
- Contractions were expanded into complete words.
- Punctuation, numbers and special characters, including emojis and emoticons, were removed using simple rules.
- The normalized text became simpler for further processing.

**Information Lost**
Normalization may remove useful information such as emojis, punctuation, numbers and user mentions. For example, an emoji may express the sentiment of a tweet, but removing it can reduce the emotional information.

**Result**
The text normalization pipeline was successfully implemented on 100 tweets. Lowercasing, contraction expansion, punctuation handling, and number/emoji handling were performed, and the original and normalized texts were compared.

**Learning Outcome**
I learned how to normalize raw social media text and understood that normalization makes text consistent but may also remove useful information.